### PR TITLE
add password encryption and validation methods

### PR DESCRIPTION
**Pull request by kenneth051**

### CODE REVIEW

File: app/controllers/user_controller.rb [hunk 1]
```diff
@@ -6,12 +6,17 @@ def index
...
+      return
+    end
     @new_user = User.new(user_params)
 
     if @new_user.valid?
       @new_user.save
-
-      render json: @new_user, status: 201
+      user = { email: @new_user.email, id: @new_user.id, first_name: @new_user.first_name, last_name: @new_user.last_name, user_type: @new_user.user_type }
+      render json: user, status: 201
```
Comment (naush) [inline app/controllers/user_controller.rb]: A couple of other ways to serialize model:

https://api.rubyonrails.org/classes/ActiveModel/Serializers/JSON.html#method-i-as_json
https://api.rubyonrails.org/classes/ActiveModel/Serialization.html

**Review by czarjulius — APPROVED** [review]

**Review by chokonaira — APPROVED** [review]
LGTM

File: app/controllers/user_controller.rb [hunk 2]
```diff
@@ -20,9 +25,15 @@ def signup
   def update_user
     current_user = User.find(params[:id])
     payload = JSON.parse(request.body.read)
+    error_messages = Validation.update_user_validate(payload)
+    if !error_messages.empty?
+      render json: { "errors": error_messages }, status: 400
+      return
+    end
```
Comment (naush) [inline app/controllers/user_controller.rb]: Another way to do this is to create your custom validation error and use [rescue_from](https://api.rubyonrails.org/classes/ActiveSupport/Rescuable/ClassMethods.html).

```rb
# controllers/application_controller.rb
class ApplicationController < ActionController::Base
  rescue_from CustomValidationError, with: :render_validation_error

  def render_validatoin_error(exception)
    render json: { "errors": exception.message }, status: 400
  end
end
```

This way you don't need to check messages or handle rendering errors here.

File: app/controllers/user_controller.rb [hunk 3]
```diff
@@ -20,9 +25,15 @@ def signup
   def update_user
     current_user = User.find(params[:id])
     payload = JSON.parse(request.body.read)
+    error_messages = Validation.update_user_validate(payload)
+    if !error_messages.empty?
```
Comment (naush) [inline app/controllers/user_controller.rb]: You can switch to `if error_messages.present?` to remove the negation. Make the code slightly more readable.

File: db/schema.rb [hunk 4]
```diff
@@ -23,14 +23,19 @@
...
+  end
+
   create_table "users", force: :cascade do |t|
-    t.string "first_name", null: false
-    t.string "last_name", null: false
+    t.string "first_name"
+    t.string "last_name"
     t.string "email", null: false
-    t.string "password", null: false
+    t.string "password_digest"
```
Comment (naush) [inline db/schema.rb:34]: 🤟

File: lib/validation.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    def validate_length(attribute, length, name)
+      if attribute.size < length
+        return { "#{name}": "length should be #{length} characters or more" }
+      end
+      nil
+    end
+
+    def sign_up_validate(payload)
+      [
+        self.validate_blank(payload["first_name"], "FirstName"),
```
Comment (naush) [inline lib/validation.rb]: `ActiveSupport` has a convenient method called `#camelize` that can handle this for you.

```rb
"first_name".camelize #=> FirstName
```

File: lib/validation.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
+require "active_support/core_ext"
+
+class Validation
+  class << self
+    def validate_blank(attribute, name)
+      if attribute.blank?
+        return { "#{name}": "Cannot be empty" }
+      end
+      nil
```
Comment (naush) [inline lib/validation.rb]: What I might do instead is to have a shared `errors` variable that I push error message into when the condition is met. So the code would look like:

```rb
errors << { "#{name}": "Cannot be empty" } if attribute.blank?
```

And I won't have to worry about dealing with `nil`.

**Review by naush — APPROVED** [review]
Nice security upgrade! 🔒 

Left a few suggestions on how to make the code more Rails-y.

File: config/database.yml [hunk 7]
```diff
@@ -10,7 +10,7 @@ default: &default
   timeout: 5000
   encoding: unicode
   username: postgres
-  password: psql
+  password: chaos
```
Comment (naush) [inline config/database.yml]: 🙈

Comment (kenneth051) [reply config/database.yml]: hahahaha... oh now you know my password

**Comment by naush** [comment]
LGTM

**Pull request closed without merging**